Fl command line tool

fangylo opened the pull request:
## Background

This PR was originally motivated by [this issue](https://github.com/twinstrandbio/duplex-seq/issues/471#issue-351762260). In order to collate all the pdfs, we need to run the command line tool [Ghostscript](https://www.ghostscript.com/), and do the following:
> - Create a shell command task wrapping gs
> - Only run the task if gs is available on the system path

`Fgbio` has [`Rscript`](https://github.com/fulcrumgenomics/fgbio/blob/master/src/main/scala/com/fulcrumgenomics/util/Rscript.scala) that achieves the same functionality for `R`. To reduce code duplication here we'd like to make a general `CommandLineTool` with the ability to extend to different executables such as R, GS, python, etc.

Nils suggested instead of putting this general `CommandLineTool` into `fgbio`, we should put it into `commons` and deprecate the original [Rscript]((https://github.com/fulcrumgenomics/fgbio/blob/master/src/main/scala/com/fulcrumgenomics/util/Rscript.scala)) in `fgbio`. 

## Changes made in this PR

1. Created a `CommandLineTool` that has some basic functionality to run a command line executable and the possibility to be extended:
	- Check if the executable is available
	- Check if certain modules is installed for the executable, e.g. R or python modules.
	- Run the executable if it is available.

2. Addressed the issue raised [here](https://github.com/fulcrumgenomics/fgbio/issues/397) and had `exec` and `execIfAvailable` both return `Unit` instead of `Try[Unit]`.

3. Currently [`Rscript.Available`](https://github.com/fulcrumgenomics/fgbio/blob/f05508919b647cc086dfc954e6192365e6d58b36/src/main/scala/com/fulcrumgenomics/util/Rscript.scala#L46) only returns **TRUE** when *both Rscript is available AND ggplot2 is installed*. This is a bit confusing so a separate `Rscript.ggplot2Available` is created to check the above instead. We can later deprecate the original `Rscript.Available`.

fangylo committed:
Added scala class for running command line tools

fangylo committed:
Save output to stream instead of temp file

Clean up

Fix all python module import

Minor clean up

fangylo committed:
Return `Unit` instead of `Try[Unit]` for exec and execIfAvailable

Added scaladoc & addressed comments

clintval commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,162 @@
+  /** The name of the executable such as Rscript or gs. */
+  val Executable: String
+
+  /** Command used for the above executable. */
```
Comment could be more informative. How so? Maybe provide an example in the scaladoc?

See the `{{{ }}}` notation in scaladoc:

```scala
/** Provides classes for dealing with complex numbers.  Also provides
 *  implicits for converting to and from `Int`.
 *
 *  ==Overview==
 *  The main class to use is [[my.package.complex.Complex]], as so
 *  {{{
 *  scala> val complex = Complex(4,3)
 *  complex: my.package.complex.Complex = 4 + 3i
 *  }}}
 *
 *  If you include [[my.package.complex.ComplexConversions]], you can
 *  convert numbers more directly
 *  {{{
 *  scala> import my.package.complex.ComplexConversions._
 *  scala> val complex = 4 + 3.i
 *  complex: my.package.complex.Complex = 4 + 3i
 *  }}}
 */
package complex {}
```

fangylo replied:
Trying to mimic the docs but having trouble importing the `CommandLineTool` package in `scala` shell :/

clintval replied:
Not sure I follow. You mean the interactive console?

```bash
$ mill -i _.console
```

clintval commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,162 @@
+    override def getMessage: String = s"$Executable failed with exit code $status."
+  }
+
+  /** Returns a tuple of (Boolean, String): first element is true when command can execute
```
Do not return mixed types. If you must return a heterogenous collection of data, use a `case class` instead.

The abstraction you are trying to emulate is one of `Try[String]`.

clintval commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,162 @@
+    try {
+      // Read redirected stdout to a string
+      val process = new ProcessBuilder(command: _*).redirectErrorStream(true).start()
+      val stdout  = mutable.ListBuffer[String]()
```
I'm pretty certain their are no immutable List Buffers. You can import the class directly:

```scala
import scala.collection.mutable.ListBuffer
```

clintval commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,162 @@
+  val Executable: String
+
+  /** Command used for the above executable. */
+  val TestCommand: Seq[String]
```
Shouldn't this be `TestArgs`?

clintval commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,162 @@
+
+  /** Returns true if the tool is available and false otherwise. */
+  lazy val Available: Boolean = {
+    execArgs(Executable +: TestCommand: _*)._1
```
If `execArgs` returns a `Try[String]` then this lazy trait constant becomes:

```suggestion
    execArgs(Executable +: TestArgs: _*).isSuccess
```

Which is more readable.

clintval commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,162 @@
+  }
+}
+
+trait CanRunScript {
```
Name could use some work. Also add scaladoc for what this mixin is for.

fangylo replied:
`CanRunFromScript`? `Scriptable`? Any other idea?

clintval replied:
`ScriptRunner`?

clintval commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,162 @@
+}
+
+trait CanRunScript {
+  self: CommandLineTool =>
```
I think the self-typing makes sense here!

Which requires the user to do:

```scala
abstract class Rscript extends Versioned with CommandLineTool
```

Whereas with just sub-typing, the user _could_ be allowed to do:

```scala
abstract class Rscript extends Versioned
```

clintval commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,162 @@
+trait CanRunScript {
+  self: CommandLineTool =>
+  /** Suffix for scripts that can be run by this tool. */
+  val Suffix: String
```
```suggestion
  val Suffix: FilenameSuffix
```

clintval replied:
Still needs fixing

clintval commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,162 @@
+  def execIfAvailable(script: Path, args: String*): Unit =
+    try {
+      if (Available) exec(script, args:_*)
+    } catch {case e: Exception => throw new Exception(s"Cannot execute script from $script with args: $args") }
```
Style either way but not both:

```scala
try { if (Available) exec(script, args:_*) }
catch { case e: Exception => throw new Exception(s"Cannot execute script from $script with args: $args") }
```

Or

```scala
try {
  if (Available) exec(script, args:_*)
} catch {
  case e: Exception => throw new Exception(s"Cannot execute script from $script with args: $args")
}
```

clintval commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,162 @@
+      if (Available) exec(scriptResource, args:_*)
+    } catch {case e: Exception => throw new Exception(s"Cannot execute script from $scriptResource with args:$args") }
+
+  /** Executes from a script stored at a Path if the tested executable is available. */
```
Use the imperative mood:

```scala
/** Execute a script if this executable is available, raise an exception otherwise. */
```

Here and throughout

fangylo replied:
👍 BTW just to be clear.. we aren't raising exception at this line for [`execIfAvailable`](https://github.com/fangylo/commons/blob/c06e2da1ea0acfc291914af77e47947600c78856/src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala#L77-L82) if the executable is *not* available.

clintval commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,162 @@
+      if (Available) exec(scriptResource, args:_*)
+    } catch {case e: Exception => throw new Exception(s"Cannot execute script from $scriptResource with args:$args") }
+
+  /** Executes from a script stored at a Path if the tested executable is available. */
```
Add scaladoc to indicate you are throwing an exception in one of your codepaths and why. Here and throughout

Use `@throws`

https://docs.scala-lang.org/overviews/scaladoc/for-library-authors.html#tags

clintval commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,162 @@
+  def exec(scriptResource: String, args: String*): Unit =
+    Try { writeResourceToTempFile(scriptResource) }.map(path => exec(path, args:_*))
+
+  /** Executes from a script stored at a Path. */
```
Does not scan

fangylo replied:
? not sure I know what you mean

clintval replied:
Sorry, "Does not scan" is an expression I've picked up.

It's shorthand to indicate that I can't make it through the documentation cleanly such as when there is something confusing about either the word choice, sentence structure, or definition you are trying to evoke to the reader.

clintval commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,162 @@
+
+  /** Executes a script from the classpath. */
+  def exec(scriptResource: String, args: String*): Unit =
+    Try { writeResourceToTempFile(scriptResource) }.map(path => exec(path, args:_*))
```
Why try on the write resource to temp file?

fangylo replied:
Not sure?
Changing it to:

```
  def exec(scriptResource: String, args: String*): Unit = {
    exec(PathUtil.pathTo(scriptResource), args:_*)
  }
```

clintval commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,162 @@
+  val TestCommand: Seq[String] = Seq(VersionFlag)
+
+  /** Returns version of the tool */
+  lazy val Version: String = execArgs(Executable +: TestCommand:_*)._2
```
This should be return typed as a `Try[String]`

clintval commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,162 @@
+  def TestModuleCommand(module: String): Seq[String]
+
+  /** Command used to test if multiple modules are included with the tested executable. */
+  def TestModuleCommand(modules: Seq[String]): Seq[Seq[String]] = modules.map(TestModuleCommand)
```
Why return a `Seq[Seq[String]]`? How is this method used? This does no "testing" itself.

Should these methods be private?

fangylo replied:
Probably not as this is used by `Rscript`? like here:https://github.com/fangylo/commons/blob/c06e2da1ea0acfc291914af77e47947600c78856/src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala#L140
It is used when there are multiple modules to be tested

clintval commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,162 @@
+  def TestModuleCommand(modules: Seq[String]): Seq[Seq[String]] = modules.map(TestModuleCommand)
+
+  /** Returns true if the tested module exist with the tested executable. */
+  def IsModuleAvailable(module: String): Boolean = execArgs(TestModuleCommand(module): _*)._1
```
This ain't Java:

```suggestion
  def moduleAvailable(module: String): Boolean = execArgs(TestModuleCommand(module): _*).isSuccess
```

clintval commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,162 @@
+  self: CommandLineTool =>
+
+  /** Command used to test if a module is included with the tested executable. */
+  def TestModuleCommand(module: String): Seq[String]
```
```suggestion
  def TestModuleArgs(module: String): Seq[String]
```

Here and throughout?

clintval commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,162 @@
+  def IsModuleAvailable(module: String): Boolean = execArgs(TestModuleCommand(module): _*)._1
+
+  /** Returns true if all tested modules exist with the tested executable. */
+  def IsModuleAvailable(modules: Seq[String]): Boolean =  modules.map(IsModuleAvailable).forall(_ == true)
```
We might parallel map this so they all happen asynchronously.

Look into Scala parallel collections.

clintval commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,162 @@
+  }
+}
+
+object GhostScript extends CommandLineTool with Versioned {
```
I would say leave this out of this PR and only include vendored support for Python and Rscript.

clintval commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,162 @@
+  val Executable: String = "gs"
+}
+
+object Python3 extends CommandLineTool with Versioned with Modular with CanRunScript {
```
We should have another trait for just `Python`!

```scala
trait Python extends Versioned with Modular with CanRunScript with CommandLineTool {
  val Executable: String
  val Extension: FilenameSuffix = ".py"
  def TestModuleArgs(module: String): Seq[String] = Seq(Executable, "-c", s"import $module")
}

object Python extends Python { val Executable: String = "python" }
object Python2 extends Python { val Executable: String = "python2" }
object Python3 extends Python { val Executable: String = "python3" }
```

clintval reviewed (changes requested):
Really getting there! I think in the cases where you want to return `Try[Unit]` you should just return `Unit` and then alert the developer you may raise an exception in the scaladoc.

fangylo committed:
Check multiple modules in parallel

Some more minor changes

Some more minor changes

Suit various comments

Clean up

Clean up

Add file back

.

mjhipp commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,205 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2017 Fulcrum Genomics LLC
```
```suggestion
 * Copyright (c) 2019 Fulcrum Genomics LLC
```

mjhipp commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,205 @@
+
+      require(exitCode == 0, s"Command did not execute successfully. Exit code: $exitCode")
+
+      stdout
```
```suggestion
      yieldAndThen(stdout)(sink.safelyClose())
```

then you don't need line 68

mjhipp commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,205 @@
+  private def writeResourceToTempFile(resource: String): Path = {
+    val lines = Io.readLinesFromResource(resource).toSeq
+    val path  = Io.makeTempFile("script.", suffix = Suffix)
+    path.toFile.deleteOnExit()
```
I don't think you need to specify `deleteOnExit()`, `Io.makeTempFile()` should take care of that for you

fangylo replied:
@mjhipp I don't think it's taken care of actually.

Here's the docs for createTempFile:
 
https://docs.oracle.com/javase/7/docs/api/java/nio/file/Files.html#createTempFile(java.nio.file.Path,%20java.lang.String,%20java.lang.String,%20java.nio.file.attribute.FileAttribute... 

> Where used as a work files, the resulting file may be opened using the DELETE_ON_CLOSE option so that the file is deleted when the appropriate close method is invoked. Alternatively, a shutdown-hook, or the File.deleteOnExit() mechanism may be used to delete the file automatically.

mjhipp reviewed (approved):
Looks good!

gpratt commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,205 @@
+
+trait CommandLineTool extends LazyLogging {
+
+  /** The name of the executable such as Rscript or gs. */
```
```suggestion
  /** The name of the executable. */
```

Keep it simple

gpratt commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,205 @@
+  /** The name of the executable such as Rscript or gs. */
+  val Executable: String
+
+  /** Command used for the above executable.
```
```suggestion
  /** Command used for the above executable. */
```

Keep it simple.  Will `TestArgs` be exposed to the user directly? I like the documentation, but keep it in the most obvious for the person interacting with code to see.

gpratt commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,205 @@
+
+  /** Executes a command and returns the stdout.
+    *
+    * @param command the command to be executed
```
```suggestion
    * @param command The command to be executed.
```

gpratt commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,205 @@
+
+  /** Executes a script from the classpath if this executable is available.
+    *
+    * @param scriptResource the name of the script resource on the classpath
```
```suggestion
    * @param scriptResource The name of the script resource on the classpath.
```

Last time I'll nitpick this.  Scaladoc should be written in complete sentences with proper punctuation.

fangylo replied:
@gpratt is this what we decided at? Looking at [scaladoc](https://docs.scala-lang.org/style/scaladoc.html#classes-objects-and-traits) it does not look like the first letter is capitalized. Also not in other docs within the fgbio `Commons` code base either.
(I did notice that you made it all the same in scala docs for tasks within duplex-seq though) 
Keeping this the same to fit other documentation within this code base for now.

mjhipp replied:
not for `@param` . The main comment should be full sentences.https://docs.scala-lang.org/style/scaladoc.html

gpratt commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,205 @@
+
+  /** Executes a script from the classpath if this executable is available, raise an exception otherwise.
+    *
+    * @throws ToolException when the exit code from the called process is not zero.
```
Whats the order of the scaladoc? `@param`, `@throws` and then `@return`? Make sure its consistant though this entire block.

gpratt commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,205 @@
+
+object Python extends Python  { val Executable: String = "python" }
+object Python2 extends Python { val Executable: String = "python2" }
+object Python3 extends Python { val Executable: String = "python3" }
```
Add trailing whitespace?

gpratt reviewed (approved):
I caught a few documentation and consistency issues.  

Still don't quite understand why the 4 different ways to exec a script are needed, but I'm going to assume they are necessary.

clintval commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,205 @@
+import scala.collection.mutable.ListBuffer
+import scala.util.Try
+
+trait CommandLineTool extends LazyLogging {
```
scaladoc missing

clintval commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,205 @@
+    *
+    * @param command the command to be executed
+    */
+  def execArgs(command: String*): Try[ListBuffer[String]] = {
```
Name of function does not suit action of function. See difference in word "args" vs "command".

I would have expected to use this like:

```scala
Python3.execArgs(Python3.Executable, "-m", "json.tool")
```

If we keep the signature, the function name should be `execCommand`.

clintval replied:
So syntax would better be:

```scala
Python3.execCommand(Python3.Executable, "-m", "json.tool")
```

However, I would have expected a public definition like:

```scala
Python3.execArgs("-m", "json.tool")
```

^ MY FAVORITE

clintval replied:
Consider too:

```scala
protected def execArgs(args: String*): Try[ListBuffer[String]] = {
```

fangylo replied:
When limiting it to `protected` it cannot be accessed by `Rscript`:
```
[error] /Users/fangyinlo/fyl_repos/fulcrum-commons/src/test/scala/com/fulcrumgenomics/commons/io/CommandlineToolTest.scala:22:21: method execCommand in trait CommandLineTool cannot be accessed in object com.fulcrumgenomics.commons.io.Rscript
```
Also, `sealed` only applies to `class`

clintval replied:
Right, private would limit access to the object in which it was defined. Protected, however, would limit access to users of the object, but the self-class and sub-classes could use the member directly.

As for traits, sealing is common in Scala, see: https://www.geeksforgeeks.org/scala-sealed-trait/

fangylo replied:
> Protected, however, would limit access to users of the object, but the self-class and sub-classes could use the member directly.

Sorry, I meant to say **protected**. The protected method cannot be accessed by `Rscript` here: https://github.com/fangylo/commons/blob/ec39999936b121365c212eec51227e8ae11c9488/src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala#L185

The above was the error from protected.

I also meant to say I can't to `seal` on this method specifically

fangylo replied:
However if changed to `protected` this method still _can_ be accessed by trait `Versioned` and `Modular` such as https://github.com/fangylo/commons/blob/ec39999936b121365c212eec51227e8ae11c9488/src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala#L159 and https://github.com/fangylo/commons/blob/ec39999936b121365c212eec51227e8ae11c9488/src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala#L172

(It cannot if it were `private`)

It's the `object Rscript` that cannot access the method: https://github.com/fangylo/commons/blob/ec39999936b121365c212eec51227e8ae11c9488/src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala#L185

clintval replied:
Ok that makes sense to me, because we are self-typing and not inheriting.

clintval reviewed (changes requested):
Also read up on the diff. between `sealed` vs `protected` vs `private` and mark members accordingly.

clintval commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,205 @@
+    * scala> val TestArgs = Seq("--version")
+    * scala> val Executable = "Rscript"
+    * scala> Rscript.execArgs(Executable +: TestArgs:_*)
+    * }}}
```
Can you post the return?
```scala
    * For example, to run `Rscript --version`:
    * {{{
    * scala> import com.fulcrumgenomics.commons.io._
    * scala> val TestArgs = Seq("--version")
    * scala> val Executable = "Rscript"
    * scala> Rscript.execArgs(Executable +: TestArgs:_*)
    * <Unit>
    * }}}
```

clintval commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,205 @@
+    * @return path to the temporary file
+    * */
+  private def writeResourceToTempFile(resource: String): Path = {
+    val lines = Io.readLinesFromResource(resource).toSeq
```
```suggestion
    val lines = Io.readLinesFromResource(resource)
```

?

clintval commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,205 @@
+  }
+}
+
+trait Versioned {
```
Scaladoc, here and throughout on traits.

clintval commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,205 @@
+    * {{
+    * scala> import com.fulcrumgenomics.commons.io._
+    * scala> Rscript.IsModuleAvailable(Seq("ggplot2", "dplyr"))
+    * }}
```
What was the response?

clintval commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,205 @@
+  /** Only returns true if R exists and ggplot2 is installed */
+  override lazy val Available: Boolean = {
+    val ToolAvailable: Boolean   = execArgs(Executable +: Seq(VersionFlag):_*).isSuccess
+    Seq(ToolAvailable, ModuleAvailable("ggplot2")).forall( _ == true)
```
You should short-circuit here, instead of checking all with (`Seq.exists`)

clintval replied:
Nevermind, this is already short-circuiting, my bad!

clintval commented on src/main/scala/com/fulcrumgenomics/commons/io/CommandLineTool.scala:
```diff
@@ -0,0 +1,205 @@
+    * scala> Rscript.IsModuleAvailable(Seq("ggplot2", "dplyr"))
+    * }}
+    */
+  def ModuleAvailable(modules: Seq[String]): Boolean =  modules.par.map(ModuleAvailable).forall(_ == true)
```
Here to, short-circuit with `Seq.exists`

clintval commented on src/test/scala/com/fulcrumgenomics/commons/io/CommandlineToolTest.scala:
```diff
@@ -0,0 +1,52 @@
+package com.fulcrumgenomics.commons.io
```
Will have to clean up!

clintval committed:
Implement a Rscript.ggplot2Available() method and tidy up a bit

fangylo committed:
Suit comments

Clean up

More minor changes

fangylo committed:
Change Python from trait to class

fangylo committed:
Made Python class constructor private so only its copmanion object can create instances of it

fangylo committed:
Make python trait again so it can only be singleton
